test/Test-Stores.shared.js

```diff
@@ -1,224 +1,137 @@
-
-      const file = await this.server.datastore.create(req)
-      await this.server.datastore.remove({ file_id: file.id })
-      assert.equal(fileDeletedEvent.calledOnce, true)
-      return assert.rejects(this.server.datastore.remove({ file_id: file.id }))
+      await this.datastore.create(file)
+      return this.datastore.remove(file.id)
     })
   })
 }
 
 exports.shouldWriteUploads = function () {
   describe('write', function () {
-    it('should reject write streams that are not open yet', function () {
+    const file = new File('1234', `${this.testFileSize}`, undefined, 'filename d29ybGRfZG9taW5hdGlvbl9wbGFuLnBkZg==,is_confidential');
+
+    it('should reject write streams that can not be open', async function () {
       const stream = fs.createReadStream(this.testFilePath)
-      return this.server.datastore.write(stream, null, 0).should.be.rejectedWith(500)
+      return this.datastore.write(stream, 'doesnt_exist', 0).should.be.rejected()
     })
 
-    it('should open a stream, resolve the new offset, and emit upload complete', function (done) {
-      const uploadCompleteEvent = sinon.fake()
-      const req = {
-        headers: {
-          'upload-length': this.testFileSize.toString(),
-          'upload-metadata': 'foo bar',
-        },
-        url: this.storePath,
-      }
-
-      this.server.datastore.on(EVENTS.EVENT_UPLOAD_COMPLETE, uploadCompleteEvent)
-
+    it('should reject whean readable stream has an error', async function () {
       const stream = fs.createReadStream(this.testFilePath)
-      const size = this.testFileSize
-      let id
-
-      stream.once('open', () => {
-        this.server.datastore
-          .create(req)
-          .then((file) => {
-            id = file.id
-            return this.server.datastore.write(stream, file.id, 0)
-          })
-          .then((offset) => {
-            assert.equal(offset, size)
-            assert.equal(uploadCompleteEvent.calledOnce, true)
-            return this.server.datastore.getOffset(id)
-          })
-          .then((stats) => {
-            assert.equal(stats.upload_length, size)
-          })
-          .then(done)
-          .catch(done)
-      })
+      return this.datastore.write(stream, 'doesnt_exist', 0).should.be.rejected()
     })
+
+    it('should write a stream and resolve the new offset', async function () {
+      await this.datastore.create(file);
+
+      const readable = fs.createReadStream(this.testFilePath);
+      const offset = await this.datastore.write(readable, file.id, 0);
+      assert.equal(offset, this.testFileSize);
+    });
+
+    it('should reject when stream is destroyed', async function () {
+      await this.datastore.create(file);
+
+      const readable = new stream.Readable();
+      const offset = this.datastore.write(readable, file.id, 0);
+
+      readable.push('some data');
+      readable.destroy();
+
+      return offset.should.be.rejected();
```

Comment:
> I'm not really sure why this last test is failing. It works for me with node 14 and 16. Also the code looks like it should be rejected as expected.

I ran into this issue too with mocha sometimes, it's a bit annoying with `async`/`await`. I think what happens is that you don't await the `datastore.write`, so you may return an unresolved promise?

```suggestion
      await assert.rejects(offset, Error) // or something more specific than `Error`
```

Reply:
> @mitjap any chance you'll have time for the final tiny steps to get this merged? I'm thinking of starting with the TypeScript rewrite and such after this PR to avoid merge conflicts.
I will give this last test one more try to push this forward. If I can't fix this to work on CI I'll just comment out this test.

When you are on TypeScript rewrite please let me know if I can pitch in.